### ISSUE TITLE
Building kernel images to support Rust unit testing

### ISSUE THREAD

**Issue by dodsonmg** [rust-vmm/vmm-reference#51]
I'm trying to familiarise myself with `rust-vmm` and have started with `vmm-reference`.  After following the [getting started](https://github.com/rust-vmm/vmm-reference/blob/master/docs/getting-started.md) guide, I tried to run the Rust unit tests for the `vmm` crate (simply `cargo test`).

Several tests failed because I hadn't built kernel images in the "default" location hard-coded into the unit tests:

https://github.com/rust-vmm/vmm-reference/blob/1c5ed93058ba83b87a18052a8ef2975752932edd/src/vmm/src/lib.rs#L453-L461

These require the images to be in `/resources/kernel` and to be named `bzimage-hello-busybox-halt` (the bzimage) and `vmlinux-hello-busybox-halt` (the elf).

Examples in the [getting started](https://github.com/rust-vmm/vmm-reference/blob/master/docs/getting-started.md) guide build kernel images in `/tmp`.

The [scripts](https://github.com/rust-vmm/vmm-reference/tree/master/resources/kernel) for generating the images in `/resources/kernel` allow you to specify a working directory, so you can go ahead and specify `/resources/kernel`; however:
- The scripts place the actual image in a subdirectory (e.g., `linux-4.14.176`), and so they must still be copied into `/resources/kernel`, and
- By default, the scripts name both the bzimage and elf images `vmlinux-hello-busybox-halt`; though the `-k` option allows you to specify the desired name (e.g., `bzimage-hello-busybox`, to which `-halt` gets appended).

I recognise this is not a big deal, but I'm raising it here because I assume I'm missing something.  I'm able to run (and pass!) all the unit tests after building the images with a script like this:

```
./make_kernel_busybox_image.sh -f elf -k vmlinux-hello-busybox -w $(pwd) -j 2 -h && \
cp linux-*/vmlinux-hello-busybox-halt .

./make_kernel_busybox_image.sh -f bzimage -k bzimage-hello-busybox -w $(pwd) -j 2 -h && \
cp linux-*/bzimage-hello-busybox-halt .
```

That's sufficiently clunky that I'm probably wrong somewhere.  Is there a better way for me to be setting up and running unit and integration tests so I can have a go at some of the "good first issue" issues and not wait until I've submitted a PR before learning I've done something dumb?

**Comment by andreeaflorescu** [comment]
Thanks for opening the issue!

You can also check how the CI is doing it right now (which sounds oddly similar to what you're doing).
First, the CI builds the dependencies with this script:
- https://github.com/rust-vmm/vmm-reference/blob/master/.buildkite/deps-pipeline.yml

Then, before running the tests, it copies the files generated by the previous step where the unit tests expects them to be:
- https://github.com/rust-vmm/vmm-reference/blob/master/.buildkite/hooks/post-checkout

In other words, this is a bit broken and we will work on fixing it. We'll use this issue to track progress.

**Comment by dodsonmg** [comment]
Thanks.  That's helping me piece together how this works.  Lot's of this is new to me...

In general, if I'm wanting to iteratively test changes locally, is the 'right' way to launch a container and run `cargo test`, similar to the recommendation in `vm-virtio`?

https://github.com/rust-vmm/vm-virtio#tests

**Comment by andreeaflorescu** [comment]
@dodsonmg yap, that sounds right.

I personally don't use docker because it gets weird at times to setup the network, so I basically just install all the tooling locally and directly run `cargo test`, but this can generate clutter, and then you might end up with a different rust version than the one in the container (which would mean that style checks might fail).

**Comment by andreeaflorescu** [comment]
@dodsonmg with the PR #67 being merged, now you no longer need to build resources locally. The resources are downloaded from S3, and there is a fake local cache (meaning that if the resources are available, they'll no longer be downloaded). Should we close this issue?
